modules/tensor_mechanics/src/kernels/TotalLagrangianStressDivergence.C

```diff
@@ -0,0 +1,209 @@
+#include "TotalLagrangianStressDivergence.h"
+
+#include "HomogenizationConstraintIntegral.h"
+
+registerMooseObject("TensorMechanicsApp", TotalLagrangianStressDivergence);
+
+InputParameters
+TotalLagrangianStressDivergence::validParams()
+{
+  InputParameters params = Kernel::validParams();
+
+  params.addRequiredParam<unsigned int>("component", "Which direction this kernel acts in");
+  params.addRequiredCoupledVar("displacements", "The displacement components");
+  params.addParam<bool>("large_kinematics", false, "Use large displacement kinematics");
+
+  // This kernel requires use_displaced_mesh to be off
+  params.suppressParameter<bool>("use_displaced_mesh");
+
+  params.addCoupledVar("macro_gradient", "Optional scalar field with the macro gradient");
+
+  params.addParam<MultiMooseEnum>("constraint_types",
+                                  HomogenizationConstants::mooseConstraintType,
+                                  "Type of each constraint: strain or stress");
+
+  return params;
+}
+
+TotalLagrangianStressDivergence::TotalLagrangianStressDivergence(const InputParameters & parameters)
+  : DerivativeMaterialInterface<Kernel>(parameters),
+    _ld(getParam<bool>("large_kinematics")),
+    _component(getParam<unsigned int>("component")),
+    _ndisp(coupledComponents("displacements")),
+    _disp_nums(_ndisp),
+    _disp_vars(_ndisp),
+    _pk1(getMaterialPropertyByName<RankTwoTensor>("pk1_stress")),
+    _dpk1(getMaterialPropertyByName<RankFourTensor>("pk1_jacobian")),
+    _macro_gradient_num(isCoupledScalar("macro_gradient", 0) ? coupledScalar("macro_gradient") : 0),
```

Comment:
```suggestion
    _macro_gradient_num(isParamValid("macro_gradient") ? coupledScalar("macro_gradient") : 0),
```

Reply:
Isn't this dangerous? If `macro_gradient` isn't coupled, and another scalar variable exists in the system, then `0` might correspond to another variable -- then there might be a segfault below.

Reply:
Now fixed with the new subclass setup.